### PR TITLE
AWS S3 copy and move tasks and `S3Bucket` methods

### LINKED ISSUE #315: AWS S3 copy and move tasks and `S3Bucket` methods
Archiving and backing up files in S3 is a common operation, but transferring them over a network then transferring them back is very inefficient and unnecessary in most cases. Built-in methods for copying and moving S3 objects would be great.

# Expectation / Proposal

Create tasks like `prefect_aws.s3.s3_list_objects` to copy and move objects using the S3 API's [copy](https://docs.aws.amazon.com/AmazonS3/latest/API/API_CopyObject.html) and [delete](https://docs.aws.amazon.com/AmazonS3/latest/API/API_DeleteObject.html) operations.  Add similar methods to the `S3Bucket` class.

```python

from prefect_aws.s3 import s3_copy, s3_move

@flow
async def test_flow():
    await s3_copy(
        source_path="object",
        source_bucket_name="my-bucket",
        target_path="my-bucket/object",
        target_bucket_name="my-backup-bucket",
        aws_credentials=aws_credentials,
        aws_client_parameters=client_parameters,
    )
    await s3_move(
        source_path="object",
        source_bucket_name="my-bucket",
        target_path="archive/object",
        aws_credentials=aws_credentials,
        aws_client_parameters=client_parameters,
    )
```

```python
from prefect_aws import S3Bucket

my_bucket = S3Bucket("my-bucket")
my_bucket.copy_object("critical/object", "my-bucket/object", target_bucket_name="my-backup-bucket")

my_bucket.move_object("object", "archive/object")
```

# Traceback / Example

- [x] I would like to [help contribute](https://PrefectHQ.github.io/prefect-aws/#contributing) a pull request to resolve this!

### PR DESCRIPTION
**Pull request by dominictarro**
Added

- `prefect_aws.s3.s3_copy`
- `prefect_aws.s3.s3_move`
- `prefect_aws.s3.S3Bucket.copy_object`
- `prefect_aws.s3.S3Bucket.move_object`

Closes #315

### Example

```python
from prefect_aws.s3 import s3_copy, s3_move

@flow
async def test_flow():
    await s3_copy(
        source_path="object",
        source_bucket_name="my-bucket",
        target_path="my-bucket/object",
        target_bucket_name="my-backup-bucket",
        aws_credentials=aws_credentials,
        aws_client_parameters=client_parameters,
    )
    await s3_move(
        source_path="object",
        source_bucket_name="my-bucket",
        target_path="archive/object",
        aws_credentials=aws_credentials,
        aws_client_parameters=client_parameters,
    )
```

```python
from prefect_aws import S3Bucket

my_bucket = S3Bucket("my-bucket")
my_bucket.copy_object("critical/object", "my-bucket/object", target_bucket_name="my-backup-bucket")

my_bucket.move_object("object", "archive/object")
```

### Checklist

- [x] References any related issue by including "Closes #<Issue Number>" or "Closes <Issue URL>".
  - If no issue exists and your change is not a small fix, please [create an issue](https://github.com/PrefectHQ/prefect-aws/issues/new/choose) first.
- [x] Includes tests or only affects documentation.
- [x] Passes `pre-commit` checks.
  - Run `pre-commit install && pre-commit run --all` locally for formatting and linting.
- [ ] Includes screenshots of documentation updates.
  - Run `mkdocs serve` view documentation locally.
- [x] Summarizes PR's changes in [CHANGELOG.md](https://github.com/PrefectHQ/prefect-aws/blob/main/CHANGELOG.md)

### CODE REVIEW

File: prefect_aws/s3.py [hunk 1]
```diff
@@ -147,6 +147,163 @@ async def example_s3_upload_flow():
...
+
+        example_copy_flow()
+        ```
+
+    """
+    logger = get_run_logger()
+
+    s3_client = aws_credentials.get_boto3_session().client(
+        "s3", **aws_client_parameters.get_params_override()
+    )
```
Comment (desertaxle) [inline prefect_aws/s3.py:231]: I'd recommend using `get_client` to ensure that the client parameters on `aws_credentials` are respected.

```suggestion
    s3_client = aws_credentials.get_client("s3")
```

File: tests/test_s3.py [hunk 2]
```diff
@@ -811,3 +985,76 @@ def test_upload_from_folder(
                 break
         else:
             raise AssertionError("Files did upload")
+
+    @pytest.mark.parametrize("client_parameters", aws_clients[-1:], indirect=True)
+    def test_copy_object(
+        self,
+        s3_bucket_with_object: S3Bucket,
+        s3_bucket_2_empty: S3Bucket,
+        client_parameters,
```
Comment (desertaxle) [inline tests/test_s3.py:994]: It looks like the `client_parameter` fixture isn't used in this test.

File: tests/test_s3.py [hunk 3]
```diff
@@ -811,3 +985,76 @@ def test_upload_from_folder(
...
+        assert s3_bucket_with_object.read_path("object_copy_1") == b"TEST"
+
+        with pytest.raises(ClientError):
+            assert s3_bucket_with_object.read_path("object") == b"TEST"
+
+    @pytest.mark.parametrize("client_parameters", aws_clients[-1:], indirect=True)
+    def test_move_object_to_nonexistent_bucket_fails(
+        self,
+        s3_bucket_with_object: S3Bucket,
+        client_parameters,
```
Comment (desertaxle) [inline tests/test_s3.py:1029]: It looks like the `client_parameter` fixture isn't used in this test. Can it be removed?

File: tests/test_s3.py [hunk 4]
```diff
@@ -811,3 +985,76 @@ def test_upload_from_folder(
...
+        with pytest.raises(ClientError):
+            s3_bucket_with_object.move_object("object", "object")
+        assert s3_bucket_with_object.read_path("object") == b"TEST"
+
+    @pytest.mark.parametrize("client_parameters", aws_clients[-1:], indirect=True)
+    def test_move_object_between_buckets(
+        self,
+        s3_bucket_with_object: S3Bucket,
+        s3_bucket_2_empty: S3Bucket,
+        client_parameters,
```
Comment (desertaxle) [inline tests/test_s3.py:1052]: It looks like the `client_parameter` fixture isn't used in this test. Can it be removed?

**Review by desertaxle — CHANGES_REQUESTED** [review]
Thanks for the contribution! Let me know if you have any questions about any of my comments!

File: prefect_aws/s3.py [hunk 5]
```diff
@@ -1023,3 +1180,163 @@ async def upload_from_folder(
...
+    async def copy_object(
+        self,
+        from_path: Union[str, Path],
+        to_path: Union[str, Path],
+        to_bucket: Optional[Union["S3Bucket", str]] = None,
+        **copy_kwargs,
+    ) -> str:
+        """Uses S3's internal
+        [CopyObject](https://docs.aws.amazon.com/AmazonS3/latest/API/API_CopyObject.html)
+        to copy objects within or between buckets.
```
[the remark this answers is not in the document]
Comment (dominictarro) [reply prefect_aws/s3.py:1194]: Yeah so the `stream_from` is good when there isn't a single set of credentials that can read from the source and write to the destination. The destination I gave in #276 was with a public bucket like NOAA's GHCN-D bucket, which must be read via an unsigned request. Its shortcoming is that the object has to be downloaded to the Python process, then uploaded again. Sometimes that is necessary, but sometimes we can just use S3's internal copy operation, which completes large copies in seconds.

Will add context to the docstrings of both methods.

**Comment by dominictarro** [comment]
@desertaxle 

- Added documentation to suggest when the native S3 operations should be used versus `S3Bucket.stream_from`.
- client credentials are no longer passed directly to `s3_copy` nor `s3_move`
- `s3_copy` and `s3_move` use `AwsCredentials.get_s3_client`
- `client_parameters` direct fixture removed from test functions

File: tests/test_s3.py [hunk 6]
```diff
@@ -811,3 +975,71 @@ def test_upload_from_folder(
                 break
         else:
             raise AssertionError("Files did upload")
+
+    @pytest.mark.parametrize("client_parameters", aws_clients[-1:], indirect=True)
```
Comment (desertaxle) [inline tests/test_s3.py:979]: This fixture is unused and can be removed.
```suggestion
```

File: tests/test_s3.py [hunk 7]
```diff
@@ -205,6 +224,151 @@ async def test_flow():
...
+    assert read(bucket_2, "object_copy_2") == b"TEST"
+
+    with pytest.raises(ClientError):
+        read(bucket, "object")
+
+    with pytest.raises(ClientError):
+        read(bucket, "subfolder/object_copy")
+
+
+@pytest.mark.parametrize("client_parameters", aws_clients, indirect=True)
```
Comment (desertaxle) [inline tests/test_s3.py:297]: This fixture is unused and can be removed.
```suggestion
```

**Review by desertaxle — CHANGES_REQUESTED** [review]
Thanks for making the requested update @dominictarro! I left some comments on some fixtures that aren't being used, but once those are removed this PR is good to go!

**Comment by dominictarro** [comment]
The fixture is necessary for other fixtures. Removing it throws an error like so

```
/Users/dominictarro/Documents/GitHub/prefect-aws/tests/test_s3.py::test_s3_copy failed with error: Test failed with exception
request = <SubRequest 'client_parameters' for <Function test_s3_copy>>
    @pytest.fixture
    def client_parameters(request):
>       client_parameters = request.param
E       AttributeError: 'SubRequest' object has no attribute 'param'
tests/test_s3.py:41: AttributeError
```

**Comment by desertaxle** [comment]
Ah, that's right. I forgot about the logic that populates `client_parameters`. This looks good to me then!

**Review by desertaxle — APPROVED** [review]

**Pull request merged**